File: frontend/src/types/history.ts
import type { Polygon } from "geojson";
import type { ChangeType, DetectionObject } from "./detection";
import type { ObjectCategory } from "./sheet";

/**
 * 처리 이력 (ReviewHistory) — PROJECT_BRIEF.md §3.6
 * Undo/Redo 의 before/after 단위로도 사용.
 */
export type HistoryAction =
  | "classify"
  | "edit_geometry"
  | "edit_meta"
  | "create"
  | "delete"
  | "restore";

export interface ReviewHistory {
  id: string;
  object_id: string;
  sheet_code: string;
  /** 본 이력이 속한 프로젝트. legacy 행은 null — sheet_code fallback 으로 노출됨. */
  task_id: string | null;
  model: ObjectCategory;
  change_type: ChangeType;
  geometry: Polygon; // 변경 시점 형상

  action: HistoryAction;
  before: Partial<DetectionObject> | null;
  after: Partial<DetectionObject> | null;

  reviewer: string;
  reviewed_at: string;
  memo: string | null;
}

File: backend/app/schemas/history.py
"""처리 이력 스키마."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict

from app.schemas.common import GeoJsonPolygon
from app.schemas.detection import ChangeType
from app.schemas.sheet import ObjectCategory

HistoryAction = Literal[
    "classify",
    "edit_geometry",
    "edit_meta",
    "create",
    "delete",
    "restore",
]


class ReviewHistory(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    object_id: str
    sheet_code: str
    # 본 이력이 속한 프로젝트(=task). legacy 행은 None — sheet_code fallback 으로 노출.
    task_id: str | None = None
    model: ObjectCategory
    change_type: ChangeType
    geometry: GeoJsonPolygon

    action: HistoryAction
    before: dict[str, Any] | None = None
    after: dict[str, Any] | None = None

    reviewer: str
    reviewed_at: datetime
    memo: str | None = None
